source/extensions/filters/http/stateful_session/stateful_session.cc

```diff
@@ -83,6 +88,22 @@ Http::FilterHeadersStatus StatefulSession::encodeHeaders(Http::ResponseHeaderMap
     auto host = upstream_info->upstreamHost();
     if (host != nullptr) {
       session_state_->onUpdate(host->address()->asStringView(), headers);
+      if (override_attempted_ && !accounted_) {
+        // If strict=false and override host was not found, router would have picked a host.
+        // Count as failed_open if the selected host address differs from attempted override.
+        const absl::string_view selected = host->address()->asStringView();
+        if (override_address_.has_value() && selected != override_address_.value()) {
+          if (!config_->isStrict()) {
+            markFailedOpen();
+          } else {
+            // In strict mode, differing selected host should not happen; strict would 503.
+          }
+        } else {
+          // Either matched override or no mismatch observable; consider routed.
+          markRouted();
+        }
+        accounted_ = true;
+      }
```

Comment:
It would be better to update the `onUpdate` to return a bool to tell whether the selected host is different. It would be more clear and easy to follow the code (and have better performance)